packages/keystone/rateLimiting/plugin.js

```diff
@@ -0,0 +1,196 @@
+     * @param {import('./request.utils').FieldInfo} mutation
+     * @return {number}
+     */
+    #getMutationComplexity (mutation) {
+        // TODO: Obtain custom override values later if needed
+        return this.#mutationWeight
+    }
+
+    serverWillStart (service) {
+        const { schema } = service
+        this.#queriesArgList = extractPossibleArgsFromSchemaQueries(schema)
+    }
+
+    requestDidStart () {
+        return {
+            didResolveOperation: async (requestContext) => {
+                const { mutations, queries } = extractQueriesAndMutationsFromRequest(requestContext)
+
+                /** @type {Record<string, number>} */
+                const queriesComplexity = {}
+                /** @type {Record<string, number>} */
+                const mutationsComplexity = {}
+
+                for (const query of queries) {
+                    // Not include introspection queries in complexity.
+                    // Since they're disabled in production and not the part of original request
+                    if (query.name.startsWith('__')) {
+                        continue
+                    }
+                    const complexity = this.#getQueryComplexity(query)
+                    // TODO: Check override query quota if needed
+                    queriesComplexity[query.name] ??= 0
+                    queriesComplexity[query.name] += complexity
+                }
+
+                for (const mutation of mutations) {
+                    const complexity = this.#getMutationComplexity(mutation)
+                    // TODO: Check override mutation quota if needed
+                    mutationsComplexity[mutation.name] ??= 0
+                    mutationsComplexity[mutation.name] += complexity
+                }
+
+                const allQueriesComplexity = Object.values(queriesComplexity).reduce((acc, curr) => acc + curr, 0)
+                const allMutationsComplexity = Object.values(mutationsComplexity).reduce((acc, curr) => acc + curr, 0)
+                const requestComplexity = allQueriesComplexity + allMutationsComplexity
+
+                requestContext.context.req.complexity = {
+                    details: {
+                        queries: queriesComplexity,
+                        mutations: mutationsComplexity,
+                    },
+                    queries: allQueriesComplexity,
+                    mutations: allMutationsComplexity,
+                    total: requestComplexity,
+                }
+
+                const { isAuthed, key } = extractQuotaKeyFromRequest(requestContext)
+                const maxQuota = isAuthed ? this.#authedQuota : this.#nonAuthedQuota
+
+                /** @type {string | null} */
+                const currentValue = await this.#redisClient.get(key)
+                const usedQuota = parseInt(currentValue) || 0
+
+                if (usedQuota + requestComplexity > maxQuota) {
+                    // TODO: Throw error instead here
```

Comment:
Maybe add some log or metric hit instead of error for now?

Reply:
It's already added. I've added all calculated info to `context.req`. And pino plugin will extract it later for Kibana to analyse